Handle Push Notifications in Dart, including background messages, notification taps (#166)
ably/ably-flutter: PushNotifications.md

Devices need to be [activated](#device-activation) with Ably once. Once activated, you can use their device ID, client ID or push token (APNs device token/ FCM registration token) to push messages to them using the Ably dashboard or a [Push Admin](https://ably.com/documentation/general/push/admin) (SDKs which provide push admin functionality, such as [Ably-java](https://github.com/ably/ably-java), [Ably-js](https://github.com/ably/ably-js), etc.). However, to send push notifications through Ably channels, devices need to [subscribe to a channel for push notifications](#subscribing-to-channels-for-push-notifications). Once subscribed, messages on that channel with a [push payload](#sending-messages) will be sent to devices which are subscribed to that channel.

### Ensure `WidgetsBinding` is initialized

> If calling any methods from the `ably_flutter` package in your `main` function, call `WidgetsFlutterBinding.ensureInitialized();`. 

Reviewer: ```suggestion
If you invoke any methods from the `ably_flutter` package in your `main` function, then ensure that you also call `WidgetsFlutterBinding.ensureInitialized();`. 
```

Should the `WidgetsBinding` in the header read `WidgetsFlutterBinding`? Perhaps the heading could be more descriptive?

> You only need to do this if calling the functions before calling `runApp()`.

Reviewer: Perhaps combine the two sentences for clarity, e.g.

"If you invoke any methods from the `ably_flutter` package in your `main` function before the call to `runApp()`, then ensure that you also call `WidgetsFlutterBinding.ensureInitialized();`.

Author: 407807f53a9d46b692c2cca30492e33d6f441bdd

> This is done to ensure all platform methods will be successfully received by the native Ably plugin on the host platform.

Reviewer: ```suggestion
This ensures that all platform methods will be successfully received by the native Ably plugin on the host platform.
```

### Device activation

- Create a rest or realtime client: e.g. `final realtime = ably.Realtime(options: clientOptions);`
- Activate the device for push notifications with Ably: `ablyClient.push.activate();`. This only
  needs to be done once, and will be used across all future app launches, as long as the app is not deactivated. This method will throw an AblyException if it fails.

> - The future returned by `activate` is not guaranteed to complete quickly. For example, in the cases where there is no internet connection, `activate` will wait until internet connection. Therefore, any code waiting for the future to complete should be not important, e.g. logging code.

Reviewer: ```suggestion
- The future returned by `activate` is not guaranteed to complete quickly. For example, if there is no internet connection, `activate` will wait until one is available. Therefore, code waiting for the future to complete should not be important, e.g. logging code.
```

Reviewer: It's unclear to me whether you are saying that logging code _is_ important and therefore _not_ a good candidate for this. I assume not?

Author: Clarified in fd50412e271b63b00b16edb9c852c0033606678e

  // TODO add logic to show notification based on message contents. 
  print('Opting to show the notification when the app is in the foreground.');
  return true;
});
```

Internally: On Android, the SDK does this by generating the same notification Firebase creates when the app is in the background. On iOS, the SDK does this by implementing [`userNotificationCenter(_:willPresent:withCompletionHandler:)`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter).

#### Notification taps

- To be informed when a notification tap launches the application:

```dart
ably.Push.notificationEvents.notificationTapLaunchedAppFromTerminated
    .then((remoteMessage) {
  if (remoteMessage != null) {
    print('The app was launched by the user by tapping the notification');
    print(remoteMessage.data);
  }
});
```
  
- To be informed when a notification is tapped by the user whilst app is in the foreground:

```dart
ably.Push.notificationEvents.onNotificationTap.listen((remoteMessage) {
  print('Notification was tapped: $remoteMessage');
});
```

#### Data Message / Background Notification

> - When the app is in the foreground, you can listen to messages using:

Author: This was done among other improvements in e2a1958402cc43cb7e7a049b4fb1c0dcce9c8ba4 earlier today. I just didn't push it until more recently

### Receiving Messages

> To see an example of receiving messages and handling other notification related behaviour, look at [push_notification_handlers](example/lib/push_notifications/push_notification_handlers.dart) in the example app.

Reviewer: notification related behavior -> notification-related behavior (note US spelling)

But perhaps just: "For examples of handling incoming messages and dealing with notifications, see:"

Author: e6f9a88cbd0b50150943aa94cf10affc98595e1f

#### Notification Message / Alert Push Notification

> **Android**: In the case where the app is in the background / terminated, you cannot configure / disable notification messages as they are automatically shown to the user by Firebase Messaging Android SDK. To create notifications which launch the application to a certain page (notifications which contain deep links or app links), or notifications which contain buttons / actions, images, and inline replies, you should send a data message and create a notification when the message is received. 

Reviewer: In the case where -> When

Author: e6f9a88cbd0b50150943aa94cf10affc98595e1f

> **iOS**: In the case where the app is in the background / terminated, ably-flutter doesn't provide the functionality to configure / extend alert notifications on iOS, and these will automatically be shown to the user.

Reviewer: In the case where -> When

Library names should be in code font throughout (`ably-flutter`)

> To create richer (more content or more interactive) local notifications once a data message is received on both Android and iOS, take a look at [awesome_notifications](https://pub.dev/packages/awesome_notifications).

Reviewer: ```suggestion
To create richer and more interactive local notifications when a data message is received on both Android and iOS, take a look at [awesome_notifications](https://pub.dev/packages/awesome_notifications).
```

Author: I initially chose the term "richer", but I don't think it's very good. I've removed it from the docs and clarified this sentence: 3e229b82b03847083da389b6ac73e779202b4375

  - Advanced usage: 
    - To do this natively on Android instead, you could use [`notificationManager.notify`](https://developer.android.com/reference/android/app/NotificationManager#notify(int,%20android.app.Notification)). 

>     - To do this natively on iOS, you can send a background message and follow [Scheduling a Notification Locally from Your App](https://developer.apple.com/documentation/usernotifications/scheduling_a_notification_locally_from_your_app). However, on iOS, you could also [customize the appearance of an alert notification](https://developer.apple.com/documentation/usernotificationsui/customizing_the_appearance_of_notifications), by registering and implementing a [`UNNotificationContentExtension`](https://developer.apple.com/documentation/usernotificationsui/unnotificationcontentextension).

Reviewer: ```suggestion
    - To do this natively on iOS, you can send a background message and follow [Scheduling a Notification Locally from Your App](https://developer.apple.com/documentation/usernotifications/scheduling_a_notification_locally_from_your_app). On iOS, you can also [customize the appearance of an alert notification](https://developer.apple.com/documentation/usernotificationsui/customizing_the_appearance_of_notifications), by registering and implementing a [`UNNotificationContentExtension`](https://developer.apple.com/documentation/usernotificationsui/unnotificationcontentextension).
```

#### Foreground notifications

> On both Android and iOS, ably-flutter allows you to configure notification sent directly from Ably to be shown when the app is in foreground. To configure if a notification is shown in the foreground:

Reviewer: ```suggestion
On both Android and iOS, `ably-flutter` enables you to show Ably notifications when the app is in the foreground. To configure this foreground display:
```

Author: This part of the docs was outdated, I've had to update it: a4bd017f4f70b3f013bb3999ca07793c84a58968

```dart
ably.Push.notificationEvents.setOnShowNotificationInForeground((message) async {
  // TODO add logic to show notification based on message contents. 
  print('Opting to show the notification when the app is in the foreground.');
  return true;
});
```

Internally: On Android, the SDK does this by generating the same notification Firebase creates when the app is in the background. On iOS, the SDK does this by implementing [`userNotificationCenter(_:willPresent:withCompletionHandler:)`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter).

#### Notification taps

> - To be informed when a notification tap launches the app:

Reviewer: ```suggestion
- To be informed when the user taps on a notification to launch the app:
```

```dart
ably.Push.notificationEvents.notificationTapLaunchedAppFromTerminated
    .then((remoteMessage) {
  if (remoteMessage != null) {
    print('The app was launched by the user by tapping the notification');
    print(remoteMessage.data);
  }
});
```
  

> - To be informed when a notification is tapped by the user whilst the app is in the foreground:

Reviewer: ```suggestion
- To be informed when the user taps on a notification while the app is in the foreground:
```

```dart
ably.Push.notificationEvents.onNotificationTap.listen((remoteMessage) {
  print('Notification was tapped: $remoteMessage');
});
```

#### Data Message / Background Notification

- On iOS, if a notification is also present, it will be shown before `onMessage` or the callback passed to `ably.Push.notificationEvents.setOnBackgroundMessage` is called.

> - On Android, if a notification is also present, it will be shown after these methods finish (or it's Future completes).

Reviewer: it's -> its

- When the app is in the foreground, you can listen to messages using:

```dart
ably.Push.notificationEvents.onMessage.listen((remoteMessage) {
  print('Message was delivered to app while the app was in the foreground: '
      '$remoteMessage');
});
```
- This method can be synchronous or `async`.

- When the app is terminated or in the background, you can listen to messages using:

```dart
Future<void> _backgroundMessageHandler(ably.RemoteMessage message) async {
  print('Just received a background message, with:');
  print('RemoteMessage.Notification: ${message.notification}');
  print('RemoteMessage.Data: ${message.data}');
}
ably.Push.notificationEvents.setOnBackgroundMessage(_backgroundMessageHandler);
```
  - This method can be synchronous or `async`.

##### Advanced: native message handling

> Users can listen to messages in each platform using the native message listeners instead of listening to messages on the Dart side. This is not recommended unless you want to **avoid** using other plugins, such as [awesome_notifications](https://pub.dev/packages/awesome_notifications) and [flutter_local_notifications](https://pub.dev/packages/flutter_local_notifications).

Reviewer: ```suggestion
Users can listen to messages in each platform using the native message listeners instead of Dart listeners. This is not recommended unless you want to **avoid** using other plugins, such as [awesome_notifications](https://pub.dev/packages/awesome_notifications) and [flutter_local_notifications](https://pub.dev/packages/flutter_local_notifications).
```

> **Android**: This means implementing [`FirebaseMessageService`](https://firebase.google.com/docs/cloud-messaging/android/receive) and overriding `onMessageReceived` method. You must also declare the `Service` in your `AndroidManifest.xml`. Once you receive your message, you could create a richer notification, by following [Create a Notification](https://developer.android.com/training/notify-user/build-notification). As declaring this service would override the service used internal by Ably Flutter, be sure to provide Ably Flutter with your registration token. Implement the following `onNewToken` method in `FirebaseMessagingService`:

Reviewer: ```suggestion
**Android**: This requires you to implement [`FirebaseMessageService`](https://firebase.google.com/docs/cloud-messaging/android/receive) and override the `onMessageReceived` method. You must also declare the `Service` in your `AndroidManifest.xml`. Once you receive your message, you could [create a richer notification](https://developer.android.com/training/notify-user/build-notification). As declaring this service would override the service used internally by Ably Flutter, be sure to provide Ably Flutter with your registration token. Implement the following `onNewToken` method in `FirebaseMessagingService`:
```

ActivationContext.getActivationContext(this).onNewRegistrationToken(RegistrationToken.Type.FCM, registrationToken);
```

> ### My notifications do not make a sound when they are shown on the device.

Author: 53b14ed4ab49a35e6143482893c098fd66706a2b

👍

This is the troubleshooting section which is styled as an FAQ in many ways

ActivationContext.getActivationContext(this).onNewRegistrationToken(RegistrationToken.Type.FCM, registrationToken);
```

### My notifications do not make a sound when they are shown on the device

> Android: Ensure you have set up a Notification channel (which displays as notification "categories" to Android users within your application settings). The example app shows how to do this by using [flutter_local_notifications](https://pub.dev/packages/flutter_local_notifications) to configure `AndroidNotificationChannel`.

Reviewer: Could bold Android and iOS here, as you have done elsewhere.

Reviewer: I don't think Notification channel needs to be capitalized here.
